### ISSUE TITLE
Angular webpack build fails with platform specific TS files

### ISSUE THREAD

**Issue by manojdcoder** [NativeScript/nativescript-dev-webpack#625]
## Issue Checklist

- [x] You are submitting: 
- [x] You verified that this is a real problem by searching the [NativeScript Forum] and the other [issues] in this repo.
- [x] You checked our [demo apps] and the [documentation] for sample usage.

## Tell us about the problem
I have a helper plugin within my Angular project,
- ui-helper-common.ts
- ui-helper.ios.ts
- ui-helper.android.ts
- ui-helper.d.ts

After upgrading to Angular 6 + NativeScript 4.x, it throws complier error as below.
```
ui-helper/ui-helper.ts is missing from the TypeScript compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
```

## Local environment

- Platform: Mac
- Node version: v9.9.0
- NativeScript CLI version: v4.1.2

## Project data

- Platform: Both
- [Demo app](https://github.com/manojdcoder/test-app)

**Comment by NickIliev** [comment]
@manojdcoder I've created [this demo app](https://github.com/NickIliev/NS-Issues-2018-II/tree/master/nativescript-dev-webpack/issue-625) based on your  own codebase and everything works as expected on my side. Try to rebuild your own project by cleaning the platform folder

```
tns platform remove ios 

// or rm -rf platforms
```

**Comment by manojdcoder** [comment]
Thanks for testing but I'm sure the issue exists. Did you build with
bundle, aot and uglify option?

Over slack @sis0k0 confirmed me that the issue was reproduced on her end
and it's something related to latest ng tools, no workaround as of now.

**Comment by NickIliev** [comment]
@manojdcoder I've built with `--bundle` as this was provided in the issue as information -  please post the exact command you are executing

**Comment by manojdcoder** [comment]
@NickIliev Please don't use run command, it always works. Use build command, that's where the issue is - while building the IPA / APK.

**Comment by NickIliev** [comment]
The issue is reproducible with `build` command

```TS
tns build ios --bundle
```

**Comment by manojdcoder** [comment]
I was trying to debug this issue, one workaround is to append platform name to the `main` entry in the `package.json` file (may not be preferable).

For example while building for iOS, update it as
```
"main": "ui-helper.ios",
```

Another workaround I'm thinking of is, since `AngularCompilerPlugin` constructor supports `compilerOptions`, may be we can pass `paths` config with `*.${platform}` appended, but not sure what exactly the syntax is. @sis0k0 Any thoughts?

**Comment by vakrilov** [comment]
Hey @manojdcoder 
Are you still getting the same error. I've tested with [your project ](https://github.com/manojdcoder/test-app) and I get the following error when building for IOS:
```
ERROR in app/app.module.ts(21,56): error TS2339: Property 'ad' does not exist on type 'typeof "/xxxx/xxxx/xxxx/test-app/app/common/ui-helper"'.
``` 
and following error when building for Andrioid:
```
ERROR in app/app.component.ts(15,22): error TS2339: Property 'ios' does not exist on type 'typeof "/xxxx/xxxx/xxxx/test-app/app/common/ui-helper"'.
```

Are these the error you are getting also?

**Comment by manojdcoder** [comment]
@vakrilov When building for production, I'm getting this error

```
ui-helper/ui-helper.ts is missing from the TypeScript compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
```

If I make ui-helper as a external plugin, then I get the errors you mentioned above.

**Comment by jannotabamo** [comment]
any luck? I got also the issue after upgrading to webpack v0.15. the plugin `nativescript-ios-uuid`.

**Comment by manojdcoder** [comment]
@vakrilov @NickIliev Any workaround might greatly help, please let us know.

**Comment by NickIliev** [comment]
+1 with `nativescript-ios-uuid` reported [here](https://github.com/NativeScript/nativescript-dev-webpack/issues/638)

**Comment by jannotabamo** [comment]
Is there any workaround for this? might greatly help for my current project, cause I can't update to the latest version of NS and NS-webpack because of this issue. :(

**Comment by jannotabamo** [comment]
@NickIliev will the fixes for this issue will be included in NS 5.0? hoping

**Comment by manojdcoder** [comment]
Unfortunately not yet with {N} 5.0.

**Comment by jannotabamo** [comment]
@manojdcoder is there any workaround for this one? because I couldn't proceed because of this problem for months.

**Comment by manojdcoder** [comment]
@jannotabamo You may move your module to project root and mark that as an external plugin in `package.json`. You might have to manually compile the TS files inside to JS.

In my case, I had move `ui-helper` folder into project root and added that as plugin in `package.json`,

```
"tns-core-modules": "~5.0.0",
 "ui-helper": "file:ui-helper",
"zone.js": "^0.8.26"
```

Then updated _paths_ in my `tsconfig.json` as below,

```
     "paths": {
            "~/*": [
                "src/*"
            ],
            "*": [
                "./ui-helper/*",
                "./node_modules/tns-core-modules/*",
                "./node_modules/*"
            ]
        }
```

If I make any changes inside `ui-helper`, I will have to compile them manually

```
tsc ./ui-helper/*.ts 
```

**Comment by mohammadrafigh** [comment]
to add more details, the problem also exists when using platform specific components in aot build with `my-component.ios.ts` and `my-component.android.ts` and `my-component.d.ts`. the compiler fails with `Cannot determine the module for class...`.

I don't know if it's related to this issue or not.

**Comment by manojdcoder** [comment]
@NickIliev @vakrilov Any progress on this yet? It has become quite impossible to write any platform specific code within project.

**Comment by triniwiz** [comment]
A work around for this would be to create an `index.ts` with the following
```
export from './ui-helper'
```

you can check out https://github.com/triniwiz/nativescript-socketio/tree/master/src for and example

**Comment by manojdcoder** [comment]
@triniwiz I tried that already but that still doesn't solve the issue. I have few delegate implementation in the `ui-helper.ios.ts` file, on loaded event of component I may override the original delegate.

For example here is what I do with Dropdown plugin, it throws error when I compile for android. Similarly android specific code throws error on iOS compilation.

```
    onDropDownLoaded(event: EventData) {
        if (isIOS) {
            const dropDown = <DropDown>event.object,
                nativeView = <UILabel>dropDown.nativeViewProtected;

            (<any>dropDown)._dropDownDelegate = uiHelper.ios
                .AppDropDownListPickerDelegateImpl
                .initWithOriginalDelegate((<any>dropDown)._dropDownDelegate);
            nativeView.lineBreakMode = NSLineBreakMode.ByWordWrapping;
            nativeView.numberOfLines = 0;
        }
    }
```

**Comment by vtjon** [comment]
@NickIliev any updates on this? It seems like a pretty big deal. I am in the process of upgrading from NS 3.X to V5 and I cannot webpack my app. I have a delegate.ios.ts and I get errors similar to above. I also have several other platform specific TS files throughout my app.

**Comment by manojdcoder** [comment]
I kept my upgrades on hold just because of this issue. Now with no option I converted everything to a local plugin and still commenting all android specific code while running build for iOS and vice versa to avoid this error which is really annoying.

**Comment by vtjon** [comment]
fwiw, I cloned the repo and the build initially worked.  I then added delegate.ios.ts to my project and I got the error:
`ERROR in ./delegate.ts
Module build failed (from ../node_modules/@ngtools/webpack/src/index.js):
Error: /Users/jonaustin/src/NS-Issues-2018-II/nativescript-dev-webpack/issue-625/app/delegate.ts is missing from the TypeScript compilation.`

I created an ios platform specific version of app.component as well. It does not appear to throw an error on that.  The changes are here:
https://github.com/vtjon/NS-Issues-2018-II/tree/master/nativescript-dev-webpack/issue-625

**Comment by manojdcoder** [comment]
@NickIliev @vtjon That is the one issue I noticed so far, when importing platform specific files the compiler expects `.[ios|android]` extension for release build but while on `tns run` the platform extension needs to be removed.

Still running further tests.

**Comment by vtjon** [comment]
I tried tns build ios --bundle --release and still got the same error.

**Comment by manojdcoder** [comment]
@NickIliev With latest modules the compilation is far better than how it used to be, at least not a blocker any more. Thank you and team for that.

I observed the two issues below,

1. As I mentioned in my previous comment `tns run` do not like platform extensions `.[ios|android] ` in require statement. But `tns build` fails without those extensions. It will be nice to have one common solution that would work for both builds.
2. As we are specifying the [main](https://github.com/vtjon/NS-Issues-2018-II/blob/master/nativescript-dev-webpack/issue-625/app/common/package.json#L3) file explicitly in `package.json`, it should allow `import * as uiHelper from "~/common"; `. The compiler always looks for `index.ts` with this import statement, at least this used to work in earlier version.

**Comment by hosikiti** [comment]
@NickIliev 

Same issue.
If there are files with platform specific name like follows, webpack compiler loads "*.[ios/android].ts" instead of loading "*.d.ts". This behavior is inconsistent with previous one's. So we have to deal with many compile errors.

For example:
comp.android.ts
comp.ios.ts
comp.d.ts

It would be great if we can hear any progress on this.

**Comment by bagnos** [comment]
Same issue.
I have a code sharing project using schematics and I get this error using a custom plugin cloned from nativescript-version-number, using demo project all works correctly.

Module build failed (from ../node_modules/@ngtools/webpack/src/index.js):
Error: /Users/XXXXXXXXsvn/nativescript-version-number/src/test.ts is missing

**Comment by sis0k0** [comment]
This should be fixed by #843

**Comment by miroslavaivanova** [comment]
This is fixed in the current master branch and the next tag of nativescript-cli.
